text/0000-partial_types.md

```diff
@@ -0,0 +1,642 @@
+- Feature Name: `partial_types`
+- Start Date: 2023-04-28
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+# Summary
+[summary]: #summary
+
+Partial Types proposal is a generalization on "partial borrowing"-like proposals.
+
+This proposal is universal flexible tool to work **safe** with partial parameters, partial arguments, partial references and partial borrowing.
+
+Advantages: maximum type safety, maximum type control guarantee, no ambiguities, flexibility, usability and universality.
+
+
+# Motivation
+[motivation]: #motivation
+
+Safe, Flexible controllable partial parameters for functions and partial consumption (including partial not borrowing) are highly needed.
+
+Partial Types extension gives to type-checker a **mathematical guarantee** that using _simultaneously_ partial typed variable, it multiple partial references and partial borrowing is as  **safe** as using them _at a sequence_. A guarantee to borrow-checker that by & var.{x,y} or &mut var.{z,t} borrowing the whole variable and pretending to borrow just several fields is **fully safe**.
+
+And since it is a guarantee by **type**, not by **values**, it has _zero cost_ in binary! Any type error is a compiler error, so no errors in the runtime.
+
+We could apply _theoretically_ this extension to all Product Types (`PT = T1 and T2 and T3  and ...`) and it fully covers all variants of possible uses of Product Types.
+
+So, most promised candidates are Structs and Tuples.
+
+This extension is not only fully backward-compatible, but is fully forward-compatible! Forward-compatibility is an ability to use updated functions old way.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For Product Types `PT = T1 and T2 and T3  and ...` (structs, tuples) we assume they are like structs with controllable field-access.
+
+`<%permit>` - field is accessible (always implicit), `%deny` - field is forbidden to use, `%miss` - it is uninitialized field, field is forbidden to use.
```

Comment:
It is not necessary `%deny` and `%miss` fields-accesses are needed, maybe it is enough for variable be "fresh".
```rust
let p = Point{x : 1.0, y: 2.0};
let rp = & p.{x};
    // rp : & Point::{x};

*rp.y = 8.0;
// or
rp.y = &8.0;
// rp : & Point;
```
This is Ok from Type point of view, but due to internal Rust representation `p.y` in reality will be overwritten by this expression, so it must be forbidden.
Fresh initialized variables could be extended, but references, "de-reverences", moved partial variables cannot be extended.

Reply:
I understand, so this is to support initializing fields in a function, like this:

```rs
fn init(s: &MyStruct.{x, y, %miss}) -> &MyStruct {
    *s.x = 42;
    *s.y = 1337;
    s
}
```

A better name here would be `uninit` instead of `miss`. Also, I think it makes more sense to put it before the field:

```rs
s: &MyStruct.{uninit x, uninit y}
```